File: lib/ex_tinygrad/protocol.ex
defmodule ExTinygrad.Protocol do
  @moduledoc """
  Encode/decode of the XTG1 wire frame carried inside each Port `packet: 4`
  message. Mirrors `priv/worker/protocol.py`.

  Frame layout (all integers big-endian):

      4  bytes  magic "XTG1"
      8  bytes  request id (u64)
      4  bytes  JSON metadata length (u32)
      2  bytes  blob count (u16)
      2  bytes  reserved
      8*N       blob lengths (u64)
      M         UTF-8 JSON metadata
      ...       concatenated blob bytes
  """

  @magic "XTG1"

  @doc """
  Encode a frame. `meta` is a JSON-encodable map, `blobs` a list of binaries.
  Returns iodata suitable for `Port.command/2`.
  """
  @spec encode(non_neg_integer(), map(), [binary()]) :: iodata()
  def encode(req_id, meta, blobs \\ []) do
    json = JSON.encode!(meta)
    blob_count = length(blobs)

    blob_lengths =
      for blob <- blobs, into: <<>>, do: <<byte_size(blob)::unsigned-big-64>>

    header =
      <<@magic, req_id::unsigned-big-64, byte_size(json)::unsigned-big-32, blob_count::unsigned-big-16,
        0::unsigned-big-16>>

    [header, blob_lengths, json | blobs]
  end

  @doc "Decode a frame binary into `{:ok, {req_id, meta, blobs}}` or `{:error, reason}`."
  @spec decode(binary()) :: {:ok, {non_neg_integer(), map(), [binary()]}} | {:error, term()}
  def decode(
        <<@magic, req_id::unsigned-big-64, json_len::unsigned-big-32, blob_count::unsigned-big-16,
          _reserved::unsigned-big-16, rest::binary>>
      ) do
    with {:ok, blob_lengths, rest} <- take_lengths(rest, blob_count, []),
         <<json::binary-size(^json_len), rest::binary>> <- rest,
         {:ok, meta} <- decode_json(json),
         {:ok, blobs} <- take_blobs(rest, blob_lengths, []) do
      {:ok, {req_id, meta, blobs}}
    else
      {:error, _} = err -> err
      _ -> {:error, :truncated_frame}
    end
  end

  def decode(_), do: {:error, :bad_magic}

  defp take_lengths(rest, 0, acc), do: {:ok, Enum.reverse(acc), rest}

  defp take_lengths(<<len::unsigned-big-64, rest::binary>>, n, acc) when n > 0,
    do: take_lengths(rest, n - 1, [len | acc])

  defp take_lengths(_, _, _), do: {:error, :truncated_blob_lengths}

  defp take_blobs(rest, [], acc) do
    if rest == <<>>, do: {:ok, Enum.reverse(acc)}, else: {:error, :trailing_bytes}
  end

  defp take_blobs(bin, [len | rest_lens], acc) do
    case bin do
      <<blob::binary-size(^len), rest::binary>> -> take_blobs(rest, rest_lens, [blob | acc])
      _ -> {:error, :truncated_blob}
    end
  end

  defp decode_json(json) do
    {:ok, JSON.decode!(json)}
  rescue
    e -> {:error, {:invalid_json, e}}
  end
end

File: priv/worker/protocol.py
"""Framed wire protocol for the nx_tinygrad worker.

Two layers:

1. Port packet framing: the Erlang Port is opened with ``packet: 4``, so every
   message on stdin/stdout is prefixed with a 4-byte big-endian length. We read
   and write those prefixes manually (:func:`read_packet` / :func:`write_packet`).

2. The XTG1 frame carried inside each packet::

       4  bytes  magic "XTG1"
       8  bytes  request id (u64 BE)
       4  bytes  JSON metadata length (u32 BE)
       2  bytes  blob count (u16 BE)
       2  bytes  reserved
       8*N       blob lengths (u64 BE)
       M         UTF-8 JSON metadata
       ...       concatenated blob bytes

stdout is reserved for protocol frames only; all logging goes to stderr.
"""
from __future__ import annotations

import json
import struct

from errors import ProtocolError

MAGIC = b"XTG1"
_HEADER = struct.Struct(">4sQIH2x")  # magic, req_id, json_len, blob_count, 2 reserved
MAX_FRAME_BYTES = 2 * 1024 * 1024 * 1024  # 2 GiB hard bound


def read_exact(stream, n: int) -> bytes | None:
    """Read exactly ``n`` bytes, or return None on clean EOF."""
    buf = bytearray()
    while len(buf) < n:
        chunk = stream.read(n - len(buf))
        if not chunk:
            return None
        buf.extend(chunk)
    return bytes(buf)


def read_packet(stream) -> bytes | None:
    """Read one ``packet: 4`` framed message payload, or None on EOF."""
    header = read_exact(stream, 4)
    if header is None:
        return None
    (length,) = struct.unpack(">I", header)
    if length > MAX_FRAME_BYTES:
        raise ProtocolError(f"incoming frame too large: {length} bytes")
    payload = read_exact(stream, length)
    if payload is None:
        raise ProtocolError("truncated frame: EOF inside packet body")
    return payload


def write_packet(stream, payload: bytes) -> None:
    stream.write(struct.pack(">I", len(payload)))
    stream.write(payload)
    stream.flush()


def decode_frame(payload: bytes) -> tuple[int, dict, list[bytes]]:
    """Decode an XTG1 frame into ``(request_id, metadata, blobs)``."""
    if len(payload) < _HEADER.size:
        raise ProtocolError("frame shorter than header")
    magic, req_id, json_len, blob_count = _HEADER.unpack_from(payload, 0)
    if magic != MAGIC:
        raise ProtocolError(f"bad magic: {magic!r}")

    offset = _HEADER.size
    lengths_size = blob_count * 8
    if offset + lengths_size > len(payload):
        raise ProtocolError("truncated blob length table")

    blob_lengths = []
    for _ in range(blob_count):
        (blen,) = struct.unpack_from(">Q", payload, offset)
        offset += 8
        blob_lengths.append(blen)

    meta_bytes = payload[offset : offset + json_len]
    offset += json_len
    if len(meta_bytes) != json_len:
        raise ProtocolError("truncated metadata")

    try:
        meta = json.loads(meta_bytes.decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as exc:
        raise ProtocolError(f"invalid JSON metadata: {exc}") from exc
    if not isinstance(meta, dict):
        raise ProtocolError("metadata must be a JSON object")

    blobs = []
    for blen in blob_lengths:
        blob = payload[offset : offset + blen]
        offset += blen
        if len(blob) != blen:
            raise ProtocolError("truncated blob")
        blobs.append(blob)

    if offset != len(payload):
        raise ProtocolError("trailing bytes after frame")

    return req_id, meta, blobs


def encode_frame(req_id: int, meta: dict, blobs: list[bytes] | None = None) -> bytes:
    blobs = blobs or []
    meta_bytes = json.dumps(meta, separators=(",", ":")).encode("utf-8")
    parts = [
        _HEADER.pack(MAGIC, req_id, len(meta_bytes), len(blobs)),
        *[struct.pack(">Q", len(b)) for b in blobs],
        meta_bytes,
        *blobs,
    ]
    return b"".join(parts)
